### PR TITLE
Custom theme support

### PR DESCRIPTION
**Pull request by saagarjha**
Most of the stuff you'd want: system appearance matching, default themes, custom user-specified themes, procfs and documents directory integration.

### CODE REVIEW

**Review by Serena-io — APPROVED** [review]

**Review by tbodt — COMMENTED** [review]
Found a couple of bugs playing around:
- I swiped to delete a theme, the theme didn't go away, I swiped a little less and tapped the delete button, and then I got an array index crash because the theme was actually deleted the first time but the table view didn't update.
- If you tap edit, then tap a theme, then go back, the table view is no longer in editing mode (it should still be).

Comments on UI:
- Instead of having to teach people to swipe to find the duplicate button, let's put the duplicate button inside the editing screen. Also add a new theme button on the main page that duplicates the current theme.
- The dynamic type is correct on the themes page, which is jarring because it's not correct anywhere else in the settings screen. We should probably fix everything else instead of fixing this though.

File: fs/proc/ish.c [hunk 1]
```diff
@@ -14,6 +14,44 @@ bool (*get_user_default)(const char *name, char **buffer, size_t *size);
 bool (*set_user_default)(const char *name, char *buffer, size_t size);
 bool (*remove_user_default)(const char *name);
 
+static int proc_ish_show_colors(struct proc_entry *UNUSED(entry), struct proc_data *buf) {
+    proc_printf(buf,
+                "\x1B[30m" "iSH" "\x1B[39m "
```
[the remark this answers is not in the document]
Comment (saagarjha) [reply fs/proc/ish.c:21]: They're actually different; this one has carriage returns to make it work when piped into a raw tty. Plus framing to make it look pretty. It sucks to have both but I considered the complexity associated with actually running the command and decided it wasn't worth it here.

**Comment by saagarjha** [comment]
Ok, I've pushed some new changes. This should resolve the crashes when editing the theme list and the desynced "Edit" button. I've also added a "Duplicate" bar button item for each theme. However, on the main page I chose instead to have a new item appear, "Import theme", when you enter editing mode. I feel this is more useful, especially since we expose the "Duplicate" action in two other places.

File: app/TerminalView.m [hunk 2]
```diff
@@ -168,12 +168,21 @@ - (void)_updateStyle {
...
         _overrideFontSize = 0;
-    id themeInfo = @{
+    Palette *palette = prefs.palette;
+    if (self.overrideAppearance != OverrideAppearanceNone) {
+        palette = self.overrideAppearance == OverrideAppearanceLight ? prefs.theme.lightPalette : prefs.theme.darkPalette;
+    }
+    NSMutableDictionary<NSString *, id> *themeInfo = [@{
         @"fontFamily": prefs.fontFamily,
         @"fontSize": @(self.effectiveFontSize),
-        @"foregroundColor": [self cssColor:prefs.theme.foregroundColor],
```
Comment (tbodt) [inline app/TerminalView.m:174]: Can delete cssColor: now

File: app/TerminalViewController.m [hunk 3]
```diff
@@ -368,6 +368,16 @@ - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
     }
 }
 
+- (void)traitCollectionDidChange:(UITraitCollection *)previousTraitCollection {
```
Comment (tbodt) [inline app/TerminalViewController.m:371]: Why not add a traitCollectionDidChange to the various views and view controllers that need appearance updates?

File: app/Theme.m [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+    return strdup(NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES).firstObject.UTF8String);
+}
+
+@implementation UIColor (iSH)
+- (nullable instancetype)ish_initWithHexString:(NSString *)string {
+    if (!string.length) {
+        return nil;
+    }
+    NSScanner *scanner = [NSScanner scannerWithString:string];
+    // Skip the leading #
```
Comment (tbodt) [inline app/Theme.m:22]: check that it actually starts with a #?

File: app/UserPreferences.m [hunk 5]
```diff
@@ -184,9 +211,20 @@ - (instancetype)init {
...
+            // This one is a little bit special, so it needs extra handling.
+            // The backing property for this is intentionally underscored.
+            kPreferenceThemeKey: @"userTheme",
         };
 #undef property
+        
+        [self updateTheme];
+        
+        [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(updateTheme:) name:ThemesUpdatedNotification object:nil];
+        [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(updateTheme:) name:ThemeUpdatedNotification object:nil];
```
Comment (tbodt) [inline app/UserPreferences.m:227]: What's the difference between these two notifications?

File: app/UserPreferences.m [hunk 6]
```diff
@@ -133,20 +146,30 @@ - (instancetype)init {
...
+            kPreferenceBlinkCursorKey: @(NO),
+            kPreferenceCursorStyleKey: @(CursorStyleBlock),
             kPreferenceHideStatusBarKey: @(NO),
+            kPreferenceColorSchemeKey: @(ColorSchemeMatchSystem),
+            kPreferenceThemeKey: @"Default",
         }];
-        _theme = [[Theme alloc] initWithProperties:[_defaults objectForKey:kPreferenceThemeKey]];
+        // https://webkit.org/blog/10247/new-webkit-features-in-safari-13-1/
+        if (@available(iOS 13.4, *)) {
+            [_defaults registerDefaults:@{
```
Comment (tbodt) [inline app/UserPreferences.m:165]: This appears to be redundant with -approximateFont, why not always set this to kSystemMonospacedFontName and let approximateFont change that to menlo as needed?

File: app/TerminalViewController.m (hunk 3 above)
Comment (saagarjha) [reply app/TerminalViewController.m:371]: I picked something that's always in the view hierarchy.

File: app/UserPreferences.m (hunk 5 above)
Comment (saagarjha) [reply app/UserPreferences.m:227]: The former is when the list of themes changes. The latter is when the current theme changes.

File: app/UserPreferences.m (hunk 6 above)
Comment (saagarjha) [reply app/UserPreferences.m:165]: `kSystemMonospacedFontName` is morally only supposed to support ui-monospaced, which is the SF Mono font. That -approximateFont returns Menlo if you happen to set this value before 13.4 is really just there as a backstop.

**Review by Mnpn — APPROVED** [review]
UI-wise this now LGTM

**Pull request merged**